rubble/src/att/mod.rs

```diff
@@ -77,7 +77,16 @@ pub trait AttributeProvider {
     ///
     /// If `handle` does not refer to a grouping attribute, returns `None`.
     ///
-    /// TODO: Human-readable docs that explain what grouping is
+    /// Groups themselves are collections of attributes. An attribute is in
+    /// exactly zero or one groups.
+    ///
+    /// For primary services and secondary services, the BLE specification gives
+    /// exact definitions of what's in the group. The group begins with the
+    /// "primary service" or "secondary service" attribute, is followed by
+    /// the various attributes contained within that service, and ends with the
+    /// last attribute contained within that service.
+    ///
+    /// TODO: document what the BLE spec has to say about grouping for characteristics.
```

Comment:
Aren't these also "contained in the service", at least logically?

Reply:
I.... think so? I'm not entirely sure.

The bluetooth specification v4.2 describes grouping in 2.5.3 with

> Generic attribute profile defines grouping of attributes for three attribute types: «Primary Service», «Secondary Service» and «Characteristic». A group begins with a declaration, and ends as defined in Section 3.1 for services and Section 3.3 for characteristics. Not all of the grouping attributes can be used in the ATT Read By Group Type Request. The «Primary Service» and «Secondary Service» grouping types may be used in the Read By Group Type Request. The «Characteristic» grouping type shall not be used in the ATT Read By Group Type Request.

The fact that it mentions primary/secondary services and characteristics separately with regards to grouping makes me think they have some different behavior. Though, I really couldn't find (or at least, understand) any descriptions of characteristics grouping behavior, so it might not have a separate one.